### PR TITLE
[TTS] Fastpitch energy conditioning

### PR DESCRIPTION
**Pull request by subhankar-ghosh**
# What does this PR do ?

Conditions FastPitch with Energy

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by Rmich008** [comment]
B

File: nemo/collections/tts/losses/fastpitchloss.py [hunk 1]
```diff
@@ -120,6 +120,37 @@ def forward(self, pitch_predicted, pitch_tgt, len):
...
+    def output_types(self):
+        return {
+            "loss": NeuralType(elements_type=LossType()),
+        }
+
+    @typecheck()
+    def forward(self, energy_predicted, energy_tgt, len):
+        dur_mask = mask_from_lens(len, max_len=energy_tgt.size(1))
+        ldiff = energy_tgt.size(1) - energy_predicted.size(1)
+        energy_predicted = F.pad(energy_predicted, (0, ldiff, 0, 0), value=0.0)
```
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/losses/fastpitchloss.py:146]: Also curious if this may run into the case where `ldiff` is negative.

[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/losses/fastpitchloss.py:146]: I suspect the reason this was added with the pitch loss is because librosa.pyin() has a bug in which it outputs 1 more audio frame that it is supposed to when the audio length is exactly divisible by the frame step size. But in that case, the correct action would be to remove the final audio from from the ground truth rather than pad a zero to the prediction.

In every other situation, this padding would indicate there is a larger bug.

File: nemo/collections/tts/modules/fastpitch.py [hunk 2]
```diff
@@ -267,6 +282,26 @@ def forward(
...
         enc_out = enc_out + pitch_emb.transpose(1, 2)
 
+        # Predict energy
+        if self.energy_predictor is not None:
+            energy_pred = self.energy_predictor(enc_out, enc_mask).squeeze(-1)
+
+            if energy is not None:
+                if energy.shape[-1] != energy_pred.shape[-1]:
+                    # Average energy over characters
+                    energy_tgt = average_pitch(energy.unsqueeze(1), attn_hard_dur)
```
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/fastpitch.py:292]: Do we want to support "not learning alignment" anymore? I think we should remove the learn_alignment flag as we had discussed during the refactoring discussion.

File: nemo/collections/tts/modules/fastpitch.py [hunk 3]
```diff
@@ -267,6 +282,26 @@ def forward(
 
         enc_out = enc_out + pitch_emb.transpose(1, 2)
 
+        # Predict energy
+        if self.energy_predictor is not None:
+            energy_pred = self.energy_predictor(enc_out, enc_mask).squeeze(-1)
+
+            if energy is not None:
+                if energy.shape[-1] != energy_pred.shape[-1]:
```
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/fastpitch.py:290]: This condition is not needed according to our recipes, so I will just remove it.

File: nemo/collections/tts/models/fastpitch.py [hunk 4]
```diff
@@ -433,16 +481,20 @@ def validation_step(self, batch, batch_idx):
...
-        mel_loss = self.mel_loss(spect_predicted=mels_pred, spect_tgt=mels)
-        dur_loss = self.duration_loss(log_durs_predicted=log_durs_pred, durs_tgt=durs, len=text_lens)
-        pitch_loss = self.pitch_loss(pitch_predicted=pitch_pred, pitch_tgt=pitch, len=text_lens)
-        loss = mel_loss + dur_loss + pitch_loss
+        mel_loss = self.mel_loss_fn(spect_predicted=mels_pred, spect_tgt=mels)
+        dur_loss = self.duration_loss_fn(log_durs_predicted=log_durs_pred, durs_tgt=durs, len=text_lens)
+        pitch_loss = self.pitch_loss_fn(pitch_predicted=pitch_pred, pitch_tgt=pitch, len=text_lens)
+        energy_loss = 0.0
+        if energy_tgt is not None:
+            energy_loss = self.energy_loss_fn(energy_predicted=energy_pred, energy_tgt=energy_tgt, len=text_lens)
```
Comment (XuesongYang) [inline nemo/collections/tts/models/fastpitch.py:489]: Is it possible to squeeze `engergy_loss = 0.0` inside the func of `self.engergy_loss_fn` when `energy_tgt is None`? If so, we can avoid if..else.. for energy in this `FastPitchModel` class.

File: nemo/collections/tts/models/fastpitch.py [hunk 5]
```diff
@@ -433,16 +481,20 @@ def validation_step(self, batch, batch_idx):
...
+        loss = mel_loss + dur_loss + pitch_loss + energy_loss
```
Comment (XuesongYang) [inline nemo/collections/tts/models/fastpitch.py:490]: Is it possible to squeeze `engergy_loss = 0.0` inside the func of `self.engergy_loss_fn` when `energy_tgt is None`? If so, we can avoid if..else.. for energy in this `FastPitchModel` class.

**Comment by XuesongYang** [comment]
Thanks for the feature. I would suggest adding unit tests for the new `EnergyLoss` with mocking tensors, just in case all masked results are expected.

File: nemo/collections/tts/models/fastpitch.py [hunk 6]
```diff
@@ -457,8 +509,11 @@ def validation_epoch_end(self, outputs):
         self.log("val_mel_loss", mel_loss)
         self.log("val_dur_loss", dur_loss)
         self.log("val_pitch_loss", pitch_loss)
+        if outputs[0]["energy_loss"] is not None:
+            energy_loss = collect("energy_loss")
+            self.log("val_energy_loss", energy_loss)
 
-        _, _, _, _, spec_target, spec_predict = outputs[0].values()
+        _, _, _, _, _, spec_target, spec_predict = outputs[0].values()
```
Comment (XuesongYang) [inline nemo/collections/tts/models/fastpitch.py:516]: we need to be careful about the returned tuple because the order of items matters. We had bugs before and training process was blocked.

File: nemo/collections/tts/models/fastpitch.py [hunk 7]
```diff
@@ -343,10 +364,24 @@ def training_step(self, batch, batch_idx):
...
+            log_durs_pred,
+            pitch_pred,
+            attn_soft,
+            attn_logprob,
+            attn_hard,
+            attn_hard_dur,
+            pitch,
+            energy_pred,
+            energy_tgt,
+        ) = self(
```
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/models/fastpitch.py:380]: Originally the return was pretty straightforward, and more features were added later (e.g. attention). I agree that a dict would be clearer, but does it play nicely with the typechecking?

File: nemo/collections/tts/models/fastpitch.py [hunk 8]
```diff
@@ -142,16 +146,23 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         speaker_emb_condition_prosody = cfg.get("speaker_emb_condition_prosody", False)
         speaker_emb_condition_decoder = cfg.get("speaker_emb_condition_decoder", False)
         speaker_emb_condition_aligner = cfg.get("speaker_emb_condition_aligner", False)
+        energy_embedding_kernel_size = 0
+        energy_predictor = None
+        if self._cfg.get("energy_predictor", None) is not None:
+            energy_predictor = instantiate(self._cfg.energy_predictor)
+            energy_embedding_kernel_size = cfg.get("energy_embedding_kernel_size", 0)
```
Comment (XuesongYang) [inline nemo/collections/tts/models/fastpitch.py:153]: initializing `energy_embedding_kernel_size` and `energy_predictor` at line 149 and 150 is not necessary, because `self._cgf.get()` already take care of the default values if they do not exist. 

This also brings in a question: shall we move the `if...else..` to the `__init__` of `FastPitchModule` for the case when `energy_embedding_kernel_size` and `energy_predictor` do not exist? It is because you have to pass them to `FastPitchModule` no matter what values they are.

**Review by XuesongYang — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/models/fastpitch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/tts/models/fastpitch.py:380]: i see. But two variables may be still with the same `NeuralType` as checked in `@typecheck`, so typecheck may be not fully functing for this concern. Besides, our type hints have already duplicated the type check, right?

[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/tts/models/fastpitch.py:380]: we dont need to revise in this PR. I created a JIRA to trace.

[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/models/fastpitch.py:380]: At the very least `@typecheck` has caught a few dimensional bugs I've had in the past, so I'm a little loath to give that up.

Thanks for the creating a tracking issue!

File: nemo/collections/tts/models/fastpitch.py (hunk 5 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/fastpitch.py:490]: Good point, this should be doable.

File: nemo/collections/tts/models/fastpitch.py (hunk 8 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/fastpitch.py:153]: You are right about line 149. We do not need to initialize `energy_embedding_kernel_size` the `cfg.get` will take care of that. But `energy_predictor` needs to be initialized to None. Because `energy_predictor` in line 150 is variable for the Neural module for predicting energy. Whereas, `self._cfg.get("energy_predictor", None)` will give us just the config, it is only when we call `instantiate` in line 152 that `energy_predictor` gets it's value (the neural module).

Can we move the `if...else`: I don't think so. Because `energy_predictor` is the NeuralModule that we pass to FastPitchModule, what we get from cfg.get is just the config. And for consistency all initialization of Neuralmodules is done in the FastPitch init and not in FastPitchModule init.

File: nemo/collections/tts/losses/fastpitchloss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply nemo/collections/tts/losses/fastpitchloss.py:146]: I think we can remove ldiff. I have tested with a small training run, and I have not noticed any difference.

File: nemo/collections/tts/models/fastpitch.py (hunk 6 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/fastpitch.py:516]: Verified.

File: nemo/collections/tts/modules/fastpitch.py [hunk 9]
```diff
@@ -292,9 +327,11 @@ def forward(
             attn_hard,
             attn_hard_dur,
             pitch,
+            energy_pred,
+            energy_tgt,
         )
 
-    def infer(self, *, text, pitch=None, speaker=None, pace=1.0, volume=None):
+    def infer(self, *, text, pitch=None, speaker=None, energy=None, pace=1.0, volume=None):
```
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/fastpitch.py:334]: Done.

**Comment by subhankar-ghosh** [comment]
Moved PR to -> https://github.com/NVIDIA/NeMo/pull/5218

File: nemo/collections/tts/models/fastpitch.py (hunk 8 above)
Comment (XuesongYang) [reply nemo/collections/tts/models/fastpitch.py:153]: > energy_predictor needs to be initialized to None

It is still not necessary because if the input cfg is `None` then `instantiate` func will directly return `None` (ref: https://github.com/facebookresearch/hydra/blob/main/hydra/_internal/instantiate/_instantiate2.py#L178-L179). So in this case, you only need,

```python
energy_predictor = instantiate(self._cfg.get("energy_predictor", None))
```
